pkg/controller/elasticsearch/client/official.go

```diff
@@ -0,0 +1,298 @@
+		}
+		response = newResponseFrom7(res)
+	case c.es8 != nil && request.ES8 != nil:
+		res, err := request.ES8(c.es8)
+		if err != nil {
+			return err
+		}
+		response = newResponseFrom8(res)
+	default:
+		// no requests specified for the current version, nothing to do
+		return nil
+	}
+
+	if response.Body != nil {
+		defer response.Body.Close()
+	}
+	return request.ResponseHandler(response)
+}
+
+// perform performs the request using the underlying Elasticsearch clients transport
+func (c *elasticsearchClients) perform(request *http.Request) (*http.Response, error) {
+	switch {
+	case c.es6 != nil:
+		return c.es6.Transport.Perform(request)
+	case c.es7 != nil:
+		return c.es7.Transport.Perform(request)
+	default:
+		return c.es8.Transport.Perform(request)
+	}
+}
+
+// customRequest7 performs a custom request to Elasticsearch v7
+func customRequest7(ctx context.Context, es *elasticsearch7.Client, request *http.Request) (*esapi7.Response, error) {
+	request = request.WithContext(ctx)
+
+	// nolint:bodyclose
+	httpRes, err := es.Perform(request)
+	if err != nil {
+		return nil, err
+	}
+
+	res := &esapi7.Response{
+		StatusCode: httpRes.StatusCode,
+		Header:     httpRes.Header,
+		Body:       httpRes.Body,
+	}
+
+	return res, nil
+}
+
+// customRequest8 performs a custom request to Elasticsearch v8
+func customRequest8(ctx context.Context, es *elasticsearch8.Client, request *http.Request) (*esapi8.Response, error) {
+	request = request.WithContext(ctx)
+
+	// nolint:bodyclose
+	httpRes, err := es.Perform(request)
+	if err != nil {
+		return nil, err
+	}
+
+	res := &esapi8.Response{
+		StatusCode: httpRes.StatusCode,
+		Header:     httpRes.Header,
+		Body:       httpRes.Body,
+	}
+
+	return res, nil
+}
+
+// Response
+type Response struct {
+	StatusCode int
+	Header     http.Header
+	Body       io.ReadCloser
+
+	// esErrorOnce is used to ensure we only decode an error once from the body
+	esErrorOnce sync.Once
+	// esError stores the cached decoded error from the body
+	esError *officialAPIError
+}
+
+func newResponseFrom6(res *esapi6.Response) *Response {
+	return &Response{
+		StatusCode: res.StatusCode,
+		Header:     res.Header,
+		Body:       res.Body,
+	}
+}
+
+func newResponseFrom7(res *esapi7.Response) *Response {
+	return &Response{
+		StatusCode: res.StatusCode,
+		Header:     res.Header,
+		Body:       res.Body,
+	}
+}
+
+func newResponseFrom8(res *esapi8.Response) *Response {
+	return &Response{
+		StatusCode: res.StatusCode,
+		Header:     res.Header,
+		Body:       res.Body,
+	}
+}
+
+// IsError is true if the HTTP status code is outside 2xx
+func (r *Response) IsError() bool {
+	return r.StatusCode < 200 || r.StatusCode > 299
+}
+
+// ESError returns the Elasticsearch API response as an error
+func (r *Response) ESError() error {
+	r.esErrorOnce.Do(func() {
+		r.esError = newOfficialAPIError(r)
+	})
+	return r.esError
+}
+
+// officialAPIError is a non 2xx response from the Elasticsearch API
+type officialAPIError struct {
```

Comment:
`official` being used in multiple places (also in file names) feels a bit weird to me (naming 🤷‍♂).
Why not just `APIError`? If we need to "wrap" things in our own structs I guess we could move everything related to the go-elasticsearch client in its own sub-package.